app/lib/timetable/src/models/substitution.dart

```diff
@@ -0,0 +1,149 @@
+
+enum SubstitutionType {
+  /// The lesson is canceled.
+  cancelled,
+
+  /// The lesson is moved to another room.
+  placeChanged,
+
+  /// Unknown substitution type.
+  unknown,
+}
+
+sealed class Substitution {
+  final SubstitutionId id;
+  final SubstitutionType type;
+
+  /// The date of the substitution.
+  final Date date;
+
+  /// The user who created the substitution.
+  final UserId createdBy;
+
+  /// The user who updated the substitution.
+  ///
+  /// This field is only set if the substitution was updated.
+  final UserId? updatedBy;
+
+  /// Defines, if the substitution is deleted.
+  ///
+  /// We don't delete the substitution from the database. Instead, we set this
+  /// flag to `true` to mark the substitution as deleted. This is required to
+  /// send notifications to the group members when the substitution is deleted
+  /// while supporting offline mode.
+  final bool isDeleted;
+
+  const Substitution({
+    required this.id,
+    required this.type,
+    required this.date,
+    required this.createdBy,
+    this.isDeleted = false,
+    this.updatedBy,
+  });
+
+  static Substitution fromData(
+    Map<String, dynamic> data, {
+    required SubstitutionId id,
+  }) {
+    final type = SubstitutionType.values
+        .tryByName(data['type'], defaultValue: SubstitutionType.unknown);
+    final updatedBy = UserId.tryParse(data['updated']?['by']);
+    final createdBy = UserId(data['created']['by'] as String);
+    final isDeleted = data['deleted'] != null;
+    return switch (type) {
+      SubstitutionType.cancelled => SubstitutionCanceled(
+          id: id,
+          date: Date.parse(data['date'] as String),
+          createdBy: createdBy,
+          isDeleted: isDeleted,
+          updatedBy: updatedBy,
+        ),
+      SubstitutionType.placeChanged => SubstitutionPlaceChange(
+          id: id,
+          date: Date.parse(data['date'] as String),
+          createdBy: createdBy,
+          newPlace: data['newPlace'] as String,
+          isDeleted: isDeleted,
+          updatedBy: updatedBy,
+        ),
+      SubstitutionType.unknown => SubstitutionUnknown(
+          id: id,
+          date: Date.parse(data['date'] as String),
+          createdBy: createdBy,
+          isDeleted: isDeleted,
+          updatedBy: updatedBy,
+        )
+    };
+  }
+
+  Map<String, dynamic> toCreateJson({
+    required bool notifyGroupMembers,
+  }) {
+    return {
+      'type': type.name,
+      'date': date.toDateString,
+      'created': {
+        'on': FieldValue.serverTimestamp(),
+        'by': createdBy.value,
+        'notifyGroupMembers': notifyGroupMembers,
+      },
+    };
+  }
+}
+
+class SubstitutionCanceled extends Substitution {
+  const SubstitutionCanceled({
+    required super.id,
+    required super.date,
+    required super.createdBy,
+    super.isDeleted,
+    super.updatedBy,
+  }) : super(type: SubstitutionType.cancelled);
+}
+
+class SubstitutionPlaceChange extends Substitution {
```

Comment:
```suggestion
class LocationChangedSubstitution extends Substitution {
```

We keep `newPlace` though so its consistent in the db